### ISSUE TITLE
Upload of large files 4GB+ fails on Safari. UpChunk or WebKit problem?

### ISSUE THREAD

**Issue by theazgra** [muxinc/upchunk#134]
Hi,

we have noticed a problem, while uploading large video files, of size 4GB+ on MacOS Safari. I am not sure, if anything can be done in UpChunk library, or if this is solely WebKit problem.

The UpChunk dispatches `'success'` event right after the upload is started, that is because the very first call to `ChunkedStreamIterable::read()` fails, resulting in *completed* (`done`) upload.

The read call fails, because of an I/O error (NotReadableError)

![image](https://github.com/muxinc/upchunk/assets/18343952/6400b07c-3cf5-43fa-932b-10c411d1a151)

Before attempting the upload, top reports only about 100MB of free RAM, if that can mean anything.

If I truncate the file to about 3GB, the upload work.

Do you have any experience with this?

Specification:
- UpChunk v3.3.2
- Safari 17.0 (19616.1.27.211.1)
- Apple M1 with 8GB RAM
- macOS 14.0 (23A344)

**Comment by karlcow** [comment]
It might be worth to open a reduced test case on https://bugs.webkit.org/

**Comment by theazgra** [comment]
@karlcow I will do!

I have created small repro, which demonstrates, that it actually is WebKit problem.

```html
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>WebKit file stream reader</title>
</head>

<body>
    <input id="picker" type="file" />
    <label for="alternativeCheck">Use file slice</label>
    <input type="checkbox" name="alternative" id="alternativeCheck">

    <script>
        const picker = document.getElementById('picker');
        const checkbox = document.getElementById('alternativeCheck');
        
        picker.onchange = async () => {
            const file = picker.files[0];

            if (checkbox.checked) {
                // This method works
                const {done, value} = await file.slice(0, 65_536).stream().getReader().read();
                console.log(done);
                console.log(value);
            }
            else {
                // I/O operation failed
                const fileStream = file.stream();
                const fileStreamReader = fileStream.getReader();
                const {done, value} = await fileStreamReader.read()

                console.log(done);
                console.log(value);
            }
        };

    </script>
</body>
</html>
```

**Comment by theazgra** [comment]
@daytime-em I imagine, this must affect more users. Well all users using newest UpChunk versions on Safari.
Will you try to find different solution, around the broken API, or will you wait for Safari team?

**Comment by cjpillsbury** [comment]
hey @theazgra! Time permitting, we're hoping to provide an "opt in" fallback to something similar to our old implementation. Unfortunately, that solution means that those 4GB+ files will have to be **_entirely_** loaded into memory (due to the nature of the WebKit bug), which isn't great either. In the interim, I just cut a release (https://www.npmjs.com/package/@mux/upchunk/v/3.3.3) that will at least give you an `"error"` with some info that you can display to users as a stop gap if that's helpful:
```ts
this.dispatch('error', { message: `Unable to read file of size ${this.file.size} bytes. Try loading from another browser.` });
```

**Comment by theazgra** [comment]
hey @cjpillsbury , thanks for the reply and new release!

I have came up with workaround for this bug. https://github.com/theazgra/forendors-upchunk/commit/1fcd4afbe861eb52a2fe7b80251c5a9d94c24e2d

Would this, if polished, be worth of PR to MuxChunk?

Currently we, on our side, will be using `useFileSlicing` when detecting Safari.

**Comment by cjpillsbury** [comment]
Hey @theazgra glad you were able to find a workaround! I always love contributions from folks and definitely appreciate the offer, but it may be hard to work through and communicate the details of a solution on this one within the time frame we're aiming for. That said, if I don't get time to implement the solution within the next few weeks (hoping to peel away some time starting today), I'd love to work with you on a mergeable solution!

Here's the short version of what I'll be doing (at least broad strokes. Things may change as I work through the actual implementation).:
1. always attempt to use readable streams from files first
2. provide an "opt in" config variable for falling back to the old method for reading files
3. if/when an error occurs when attempting to read from the readable stream (detecting this is now implemented), fall back to the old method, aka reading the entire file into memory and slicing, similar to your implementation and also similar to the pre 3.x implementation
    1. This will be a completely different iterable implementation (might be a subclass though? I'll need to work through the details) - This is so we can easily isolate and remove the fallback if/when WebKit/Safari fixes the bug (🤞), but also to leave room for some future possible features to upchunk, like uploading directly from screen recordings instead of waiting until the recording is completed (also 🤞).

**Comment by theazgra** [comment]
Hey @cjpillsbury thanks for headsup. Looking forward to the next release, will switch to it when available. Thanks for the quick replies and fast problem solving.

I will close this issue as it is being solved in #138
